src/main/java/com/google/api/codegen/transformer/csharp/CSharpGapicSurfaceTransformer.java

```diff
@@ -0,0 +1,329 @@
+    List<StaticLangApiMethodView> apiMethods = new ArrayList<>();
+
+    for (Method method : context.getNonStreamingMethods()) {
+      MethodConfig methodConfig = context.getMethodConfig(method);
+      MethodTransformerContext methodContext = context.asMethodContext(method);
+
+      if (methodConfig.isPageStreaming()) {
+        /*if (methodConfig.isFlattening()) {
+          for (ImmutableList<Field> fields : methodConfig.getFlattening().getFlatteningGroups()) {
+            apiMethods.add(
+                apiMethodTransformer.generatePagedFlattenedMethod(methodContext, fields));
+          }
+        }
+        apiMethods.add(apiMethodTransformer.generatePagedRequestObjectMethod(methodContext));
+        apiMethods.add(apiMethodTransformer.generatePagedCallableMethod(methodContext));
+        apiMethods.add(apiMethodTransformer.generateUnpagedListCallableMethod(methodContext));*/
+      } else {
+        if (methodConfig.isFlattening()) {
+          for (ImmutableList<Field> fields : methodConfig.getFlattening().getFlatteningGroups()) {
+            apiMethods.add(transformFlattenedMethod(apiMethodTransformer.generateFlattenedMethod(methodContext, fields, ApiMethodType.FlattenedMethod)));
+            apiMethods.add(transformFlattenedMethod(apiMethodTransformer.generateFlattenedMethod(methodContext, fields, ApiMethodType.FlattenedMethodAsyncCallSettings)));
+            apiMethods.add(transformFlattenedMethod(apiMethodTransformer.generateFlattenedMethod(methodContext, fields, ApiMethodType.FlattenedMethodAsyncCancellationToken)));
+          }
+        }
+        //apiMethods.add(apiMethodTransformer.generateRequestObjectMethod(methodContext));
+        //apiMethods.add(apiMethodTransformer.generateCallableMethod(methodContext));
+      }
+    }
+
+    return apiMethods;
+  }
+  
+  private StaticLangApiMethodView transformFlattenedMethod(StaticLangApiMethodView method) {
+    switch (method.type())
+    {
+      case FlattenedMethod:
+        method = AddParam(method, "CallSettings", "callSettings", "null", "If not null, applies overrides to this RPC call.");
+        method = AddReturnDoc(method, "The RPC response.");
+        break;
+      case FlattenedMethodAsyncCallSettings:
+        method = AddParam(method, "CallSettings", "callSettings", "null", "If not null, applies overrides to this RPC call.");
+        method = MakeAsync(method);
+        method = AddReturnDoc(method, "A Task containing the RPC response.");
+        break;
+      case FlattenedMethodAsyncCancellationToken:
+        method = AddParam(method, "CancellationToken", "cancellationToken", null, "A <see cref=\"CancellationToken\"/> to use for this RPC.");
+        method = MakeAsync(method);
+        method = AddReturnDoc(method, "A Task containing the RPC response.");
+        break;
+      default:
+        throw new RuntimeException("Unexpected method type: '" + method.type() + "'");
+    }
+    return method;
```

Comment:
OK, done, but I'm not at all convinced this is better. If there's a nicer way of doing it in this way, please let me know.

In my (admitedly functional!) opinion the pattern I had used of transforming immutable objects, although somewhat ugly in Java, is nicer as it allows the common parts to be generated by common code, but then the language-specific transforms are applied afterwards. Yes, values seen in the final view model object come from two places, but it does mean the common generation code isn't confused (as it is a bit now) with language-specific concerns.
(This is a well-used pattern in more functional languages. OK, OK, I know Java is functional however much I try to convince it to be ;)